sdk/python/kfp_tekton/tekton.py

```diff
@@ -311,3 +311,39 @@ def __enter__(self) -> Union[Tuple[TektonLoopIterationNumber, LoopArguments], _f
   def enumerate(self) -> dsl.ParallelFor:
     self.call_enumerate = True
     return self
+
+
+class AddOnGroup(dsl.OpsGroup):
+    """
+    Represents an AddOn group containing ops and group of OpsGroups.
+    This class is the base class for a customized OpsGroup. Users can
+    develop their own OpsGroup. The customized OpsGroup maps to a
+    custom task in Tekton.
+    """
+    TYPE_NAME = 'addon_group'
+    TASK_TYPE = 'task'
+    RUN_TYPE = 'run'  # means custom task
+    DEFAULT_KIND = 'AddOnGroup'
+    DEFAULT_APIVERSION = 'custom.tekton.dev/v1alpha1'
+
+    def __init__(self, task_type: str = RUN_TYPE,
+                kind: str = DEFAULT_KIND,
+                api_version: str = DEFAULT_APIVERSION,
+                is_finally: bool = False,
+                parallelism: int = None,
+                params: Dict[str, Union[dsl.PipelineParam, str, int]] = {}):
```

Comment:
I can see that `params` here are the ones that are both for call/instantiation and `spec` --- but in some cases we need to separate the two, just like in range loops: `from`-`step`-`end` for call, but only one param for `spec`.

Initially I thought that maybe the `post_update` can fix that, but I can't see a way to access those spec-parameters inside of the `pipelineSpec`.

I would expect to be able to get sth like:
```python
with Handler(from, step, end) as params:
  item = params.item # or: params['item']
  exit_op_0 = PrintOp(item) 
```

Reply:
I think you can override the `__enter__` when you implement your class that extends `AddOnGroup`. just remember to call `super().__enter__`.

Reply:
@yhwang When I'm trying to do that, I'm hitting a problem with creating a parameter originating from this OpsGroup:
* if I return it as `dsl.PipelineParam('foo')`, it will get an entry in the call, which I can't remove via `post_update` (as `post_update` can only edit the `spec` part); additionally, if I use it in a downstream task, I'm hitting:
  ```
  Traceback (most recent call last):
    File "(...)/add_on_group__test.py", line 373, in <module>
      TektonCompiler().compile(pipeline_1e436c51_937d_47c8_a440_a4711559e589, __file__.replace('.py', '.yaml'))
    File "(...)/kfp_tekton/compiler/compiler.py", line 1509, in compile
      super().compile(pipeline_func, package_path, type_check, pipeline_conf=pipeline_conf)
    File "(...)/lib/python3.9/site-packages/kfp/compiler/compiler.py", line 1175, in compile
      self._create_and_write_workflow(
    File "(...)/kfp_tekton/compiler/compiler.py", line 1642, in _create_and_write_workflow
      pipeline_loop_crs, workflow = self.prepare_workflow(
    File "(...)/kfp_tekton/compiler/compiler.py", line 1574, in prepare_workflow
      workflow = self._create_workflow(
    File "(...)/kfp_tekton/compiler/compiler.py", line 1467, in _create_workflow
      workflow = self._create_pipeline_workflow(
    File "(...)/kfp_tekton/compiler/compiler.py", line 1194, in _create_pipeline_workflow
      substitute_param = '$(tasks.%s.results.%s)' % (sanitize_k8s_name(pp.op_name), sanitize_k8s_name(pp.name))
    File "(...)/kfp_tekton/compiler/_k8s_helper.py", line 46, in sanitize_k8s_name
      k8s_name = re.sub('[^-_./0-9A-Za-z]+', '-', name)
    File "(...)/lib/python3.9/re.py", line 210, in sub
      return _compile(pattern, flags).sub(repl, string, count)
  TypeError: expected string or bytes-like object
  ```
* if I return it as `dsl.PipelineParam('foo', op_name=self.name)`, I hit:
  ```
  Traceback (most recent call last):
    File "(...)/add_on_group__test.py", line 363, in <module>
      TektonCompiler().compile(pipeline_1e436c51_937d_47c8_a440_a4711559e589, __file__.replace('.py', '.yaml'))
    File "(...)/kfp_tekton/compiler/compiler.py", line 1509, in compile
      super().compile(pipeline_func, package_path, type_check, pipeline_conf=pipeline_conf)
    File "(...)/lib/python3.9/site-packages/kfp/compiler/compiler.py", line 1175, in compile
      self._create_and_write_workflow(
    File "(...)/kfp_tekton/compiler/compiler.py", line 1642, in _create_and_write_workflow
      pipeline_loop_crs, workflow = self.prepare_workflow(
    File "(...)/kfp_tekton/compiler/compiler.py", line 1574, in prepare_workflow
      workflow = self._create_workflow(
    File "(...)/kfp_tekton/compiler/compiler.py", line 1467, in _create_workflow
      workflow = self._create_pipeline_workflow(
    File "(...)/kfp_tekton/compiler/compiler.py", line 867, in _create_pipeline_workflow
      raw_templates = self._create_dag_templates(pipeline, op_transformers, params)
    File "(...)/kfp_tekton/compiler/compiler.py", line 578, in _create_dag_templates
      inputs, outputs = self._get_inputs_outputs(
    File "(...)/kfp_tekton/compiler/compiler.py", line 708, in _get_inputs_outputs
      upstream_op = pipeline.ops[param.op_name]
  KeyError: 'addon-group-1'
  ```